### PR TITLE
tunnel/tunnel.go: add SetPcapFd()

### PR DESCRIPTION
**Pull request by Lanius-collaris**
celzero/rethink-app#1309

PoC:
[Lanius-collaris/playground1](https://github.com/Lanius-collaris/playground1)
```
func GoHello(fd int32) int64{
	out:=os.NewFile(uintptr(fd), "")
	defer out.Close()
	str:=fmt.Sprintf("hello from golang, pid: %d\n", os.Getpid())
	len,_:=out.Write([]byte(str))
	return int64(len)
}
```

### CODE REVIEW

File: tunnel/tunnel.go [hunk 1]
```diff
@@ -204,6 +206,30 @@ func (t *gtunnel) SetPcap(fpcap string) error {
...
+	if fpcap < 0 {
+		log.I("netstack: pcap closed (ignored-err? %v)", ignored)
+		return nil // nothing else to do; pcap is closed
+	} else if fpcap < 3 {
+		// if fpcap is 0, 1, or 2 then pcap is written to stdout
+		ok := pcap.log(true)
+		log.I("netstack: pcap(int32 %d)/log(%t)", fpcap, ok)
+		return nil // fdbased will write to stdout
+	} else {
+		fout := os.NewFile(uintptr(fpcap), "")
```
Comment (ignoramous) [inline tunnel/tunnel.go:226]: Does this open the file in *append* mode?

**Comment by ignoramous** [comment]
Passing a SAF `fd` to `tunnel.SetPcapFd` works?

File: tunnel/tunnel.go (hunk 1 above)
Comment (Lanius-collaris) [reply tunnel/tunnel.go:226]: `os.NewFile()` doesn't open the file, it only return a `*File`, the file mode is able to be specified by the kotlin layer.
e.g.
`xxresolver.openFileDescriptor(uri, "wa")`
https://developer.android.com/reference/kotlin/android/content/ContentResolver#openFileDescriptor(android.net.Uri,%20kotlin.String)

**Comment by ignoramous** [comment]
Did this ever work? If so, we'd prefer passing FD instead. In our tests, it never did (not this particular code but the one we wrote when we first impl this feature).

**Comment by Lanius-collaris** [comment]
I have not tested on real rethink-app, but the PoC works whether I pass mode `w` or mode `wa` to `openFileDescriptor()`.
Could you tell me the commit id of your test? Maybe I can find why it doesn't work.

**Comment by ignoramous** [comment]
That code isn't public, I'm afraid. The current code (passing filepath instead of `fd`) works, so we will leave it as-is, as there are other complications with passing the fd around (such as double-close & use-after-close, since both Kotlinland & Goland may think they own the `fd`).

**Comment by Lanius-collaris** [comment]
`runtime.AddCleanup()` or older `runtime.SetFinalizer()` can be used to close the FD reliably ( in this case kotlin should use `ParcelFileDescriptor.detachFd()` ).
https://pkg.go.dev/runtime#AddCleanup

**Pull request closed without merging**